File: WebClient SDK/src/default/textRecognition/extensions/ocrSettingsUIExtension.ts
import { ArrayStorage } from 'tessa/platform/storage';
import { CardRow, FieldMapStorage } from 'tessa/cards';
import { CardTableViewControlViewModel } from '../../ui/tableViewExtension/cardTableViewControlViewModel';
import { CardTableViewRowData } from '../../ui/tableViewExtension/cardTableViewRowData';
import { CardUIExtension, ICardUIExtensionContext, IControlViewModel } from 'tessa/ui/cards';
import { clearStorageRows } from '../misc/ocrUtilities';
import { DotNetType, Guid } from 'tessa/platform';
import { OcrSettingsTypeId } from '../misc/ocrConstants';
import { reaction } from 'mobx';

/**
 * Расширение, выполняющее:
 * - очистку связанных секций при настройке маппинга полей
 * - проброс параметров маппинга в форму строк дочерних таблиц с настройками верификации.
 */
export class OcrSettingsUIExtension extends CardUIExtension {
  //#region fields

  private _disposers: (VoidFunction | null)[] | null = [];

  //#endregion

  //#region base overrides

  public shouldExecute(context: ICardUIExtensionContext): boolean {
    return Guid.equals(context.card.typeId, OcrSettingsTypeId);
  }

  public initialized(context: ICardUIExtensionContext): void {
    const cardSections = context.card.sections;
    const virtualFields = cardSections.get('OcrMappingSettingsVirtual')!.fields;
    const types = cardSections.get('OcrMappingSettingsTypes')!.rows;
    const sections = cardSections.get('OcrMappingSettingsSections')!.rows;
    const fields = cardSections.get('OcrMappingSettingsFields')!.rows;

    this.onDataChanged(types, 'TypeID', sections);
    this.onDataChanged(sections, 'SectionID', fields);

    this.selectedRowChangedHandler(
      'FieldsMappingTypesSettings',
      'TypeID',
      virtualFields,
      context.model.controls
    );
    this.selectedRowChangedHandler(
      'FieldsMappingSectionsSettings',
      'SectionID',
      virtualFields,
      context.model.controls
    );
  }

  public finalized(_context: ICardUIExtensionContext): void {
    if (this._disposers) {
      for (const disposer of this._disposers) {
        disposer?.();
      }
      this._disposers = null;
    }
  }

  //#endregion

  //#region handlers

  private onDataChanged(
    observable: ArrayStorage<CardRow>,
    observableFieldName: string,
    observer: ArrayStorage<CardRow>
  ): void {
    for (const observableRow of observable) {
      this.onFieldChanged(observableRow, observableFieldName, observer);
    }
    this.onCollectionChanged(observable, observableFieldName, observer);
  }

  private onCollectionChanged(
    observable: ArrayStorage<CardRow>,
    observableFieldName: string,
    observer: ArrayStorage<CardRow>
  ): void {
    this._disposers?.push(
      observable.collectionChanged.addWithDispose(args => {
        for (const observableRow of args.added) {
          this.onFieldChanged(observableRow, observableFieldName, observer);
        }
      })
    );
  }

  private onFieldChanged(
    observableRow: CardRow,
    observableFieldName: string,
    observer: ArrayStorage<CardRow>
  ): void {
    this._disposers?.push(
      observableRow.fieldChanged.addWithDispose(args => {
        if (args.fieldName === observableFieldName) {
          const rowId = observableRow.rowId;
          clearStorageRows(observer, r => Guid.equals(r.parentRowId, rowId));
        }
      })
    );
  }

  private selectedRowChangedHandler(
    controlName: string,
    fieldName: string,
    virtualStorage: FieldMapStorage,
    controls: ReadonlyMap<string, IControlViewModel>
  ): void {
    const control = controls.get(controlName) as CardTableViewControlViewModel;
    this._disposers?.push(
      reaction(
        () => control.selectedRow,
        () => {
          const cardRow = (control.selectedRow as CardTableViewRowData)?.cardRow;
          virtualStorage.rawSet(fieldName, cardRow?.get(fieldName), DotNetType.Guid);
        }
      )
    );
  }

  //#endregion
}

File: Source/Extensions/Tessa.Extensions.Default.Client/TextRecognition/OcrSettingsUIExtension.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Tessa.Cards;
using Tessa.Extensions.Platform.Client.UI.TableViewExtension;
using Tessa.Platform.Storage;
using Tessa.TextRecognition;
using Tessa.UI;
using Tessa.UI.Cards;
using Tessa.UI.Views.Content;

namespace Tessa.Extensions.Default.Client.TextRecognition
{
    /// <summary>
    /// Расширение, выполняющее:
    /// <para>
    ///     - очистку связанных секций при настройке маппинга полей
    /// </para>
    /// <para>
    ///     - проброс параметров маппинга в форму строк дочерних таблиц с настройками верификации.
    /// </para>
    /// </summary>
    public sealed class OcrSettingsUIExtension : CardUIExtension
    {
        #region Base Overrides

        /// <inheritdoc/>
        public override async Task Initialized(ICardUIExtensionContext context)
        {
            var cardSections = context.Card.Sections;
            var virtualFields = cardSections["OcrMappingSettingsVirtual"].RawFields;
            var types = cardSections["OcrMappingSettingsTypes"].Rows;
            var sections = cardSections["OcrMappingSettingsSections"].Rows;
            var fields = cardSections["OcrMappingSettingsFields"].Rows;

            OnDataChanged(types, "TypeID", sections);
            OnDataChanged(sections, "SectionID", fields);

            SelectedRowChangedHandler(
              "FieldsMappingTypesSettings",
              "TypeID",
              types,
              virtualFields,
              context.Model.Controls);

            SelectedRowChangedHandler(
              "FieldsMappingSectionsSettings",
              "SectionID",
              sections,
              virtualFields,
              context.Model.Controls);
        }

        #endregion

        #region Private

        private static void OnDataChanged(
            ListStorage<CardRow> observable,
            string observableFieldName,
            ListStorage<CardRow> observer)
        {
            foreach (var observableRow in observable)
            {
                OnFieldChanged(observableRow, observableFieldName, observer);
            }

            OnCollectionChanged(observable, observableFieldName, observer);
        }

        private static void OnCollectionChanged(
            ListStorage<CardRow> observable,
            string observableFieldName,
            ListStorage<CardRow> observer)
        {
            observable.ItemChanged += (_, args) =>
            {
                if (args.Item.State == CardRowState.Inserted)
                {
                    OnFieldChanged(args.Item, observableFieldName, observer);
                }
            };
        }

        private static void OnFieldChanged(
            CardRow observableRow,
            string observableFieldName,
            ListStorage<CardRow> observer)
        {
            observableRow.FieldChanged += (_, args) =>
            {
                if (args.FieldName == observableFieldName)
                {
                    var rowId = observableRow.RowID;
                    OcrHelper.ClearStorageRows(observer, r => r.ParentRowID == rowId);
                }
            };
        }

        private static void SelectedRowChangedHandler(
            string controlName,
            string fieldName,
            ListStorage<CardRow> storage,
            IDictionary<string, object> virtualStorage,
            IViewModelContainer<IControlViewModel> controls)
        {
            var control = controls[controlName] as CardTableViewControlViewModel;
            var firstRow = control.Rows.FirstOrDefault();
            if (firstRow is not null && firstRow.IsSelected)
            {
                OnRowSelected(firstRow);
            }

            control.Table.RowSelected += (_, args) =>
            {
                OnRowSelected(args.Row);
            };

            void OnRowSelected(TableRowViewModel rowModel)
            {
                var rowID = rowModel.Data.Get<Guid>(CardRow.RowIDKey);
                var row = storage.FirstOrDefault(r => r.RowID == rowID);
                virtualStorage[fieldName] = row?.Get<Guid>(fieldName);
            }
        }

        #endregion
    }
}
